Array helpers

taras opened the pull request:
- [x] Added map helper
- [x] Added reduce helper
- [x] Added filter helper
- [x] Updated README

martndemus commented on README.md:
```diff
@@ -79,8 +79,11 @@ For action helpers, this will mean better currying semantics:
   + [`underscore`](#underscore)
   + [`w`](#w)
 * [Array](#array-helpers)
+  + [`array`](#array)
```
You moved array to the top, but didn't move the section in the document itself. Either move this one back, or move the section to the correct order. I'd prefer moving this entry back to it's original place, as I don't see the need why this needs to be moved.

taras replied:
I can move it back, but I think it should be at the top because it's one of the most primitive helpers in this list. People need to know that it's available. I'm sure they'll come across it eventually but why not put it at the top so it's right there.

martndemus commented on README.md:
````diff
@@ -283,6 +286,15 @@ See also: [Ember `w` documentation](http://emberjs.com/api/classes/Ember.String.
+Maps a callback on an array.
+
+```hbs
+{{#each (map (action 'getName') users) as |fullName|}}
````
Double quotes please

martndemus commented on README.md:
````diff
@@ -446,6 +468,15 @@ Returns an array with the first `n` entries omitted.
+Reduce an array to a value.
+
+```hbs
+{{reduce (action 'sum') (array 1 2 3) 0}}
````
Double quotes please

martndemus commented on README.md:
````diff
@@ -446,6 +468,15 @@ Returns an array with the first `n` entries omitted.
+{{reduce (action 'sum') (array 1 2 3) 0}}
+```
+
+The last argument is initial value. If you omit it, undefined will be used.
````
I think it should be the second argument: `{{reduce (action "sum") 0 (array 1 2 3)}}`

martndemus commented on tests/integration/helpers/filter-test.js:
```diff
@@ -0,0 +1,73 @@
+    { foo: [1, 2, 3],  name: 'e' }
+  ]));
+
+  this.set('callback', function({ foo }) {
```
Please use `this.on` for setting actions on the context.

martndemus commented:
In summary:
- [ ] The last argument to `{{filter}}` should always be the array
- [ ] Use double quotes in HBS templates
- [ ] Use `this.on` to set actions on the test context
- [ ] Needs entries in `addon/index.js`

taras commented:
@martndemus I've done everything.

poteto commented on addon/helpers/filter.js:
```diff
@@ -0,0 +1,36 @@
+
+export default Helper.extend({
+  compute([callback, array]) {
+
```
Extra whitespace

taras replied:
@poteto sorry, I'm not sure what you mean. I checked for extra white space and I don't see it. I also looked at how `filter-by` is implemented and I don't see a difference there either. Could you please tell me what you'd like me to change about the whitespace.

martndemus replied:
I think she means delete the empty line at L13

poteto commented on addon/helpers/reduce.js:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    let cp = computed('array.[]', () => {
+      let array = this.get('array');
```
`get(this, ...)`

poteto commented on tests/integration/helpers/reduce-test.js:
```diff
@@ -0,0 +1,65 @@
+});
+
+test('It accepts a callback', function(assert) {
+  this.set('array', emberArray([ 1, 2, 3 ]));
```
No spaces within an array please, just `[1, 2, 3]` is fine

poteto commented on tests/integration/helpers/map-test.js:
```diff
@@ -0,0 +1,78 @@
+    {{~/each~}}
+  `);
+
+  this.set('callback', function({ x }) {
```
Why is this defined again?

taras replied:
This was a mistake. Thank you for catching it. I changed it to use `this.on` and `action` instead of callback.

poteto commented on tests/integration/helpers/map-test.js:
```diff
@@ -0,0 +1,78 @@
+
+  run(() => array.pushObject({ name: 'd' }));
+
+  assert.equal(this.$().text().trim(), 'abcd', 'd is added');
```
Needs a precondition test, before pushing `d`:

```js
assert.equal(this.$().text().trim(), 'abc', 'precondition');
```

poteto commented on tests/integration/helpers/filter-test.js:
```diff
@@ -0,0 +1,73 @@
+    { foo: 3, name: 'c' }
+  ]));
+
+  this.on('isOdd', ({ foo }) => foo % 2 === 1);
```
This logic is correct but it reads a little funny. `foo % 2 !== 0` feels clearer

poteto commented:
Looks good, mostly minor formatting and all tests need precondition assertions

taras commented:
@poteto I've made the changes that you suggested.

martndemus commented:
@taras you still need to fix changing `this.set` into `this.on` when setting actions on the test context.

taras commented:
@martndemus sorry, I missed a few `this.set`s. I went through all of them and replaced them with `this.on` in all of the tests that I added. Removed empty line from `map`, `reduce`, `filter` and `filter-by`.

martndemus commented on tests/integration/helpers/filter-test.js:
```diff
@@ -0,0 +1,73 @@
+  });
+
+  this.render(hbs`
+    {{~#each (filter (action 'truthyFoo') array) as |item|~}}
```
Use double quotes in HBS templates.

martndemus commented:
If you can fix the double quotes issue and then squash all your commits into one, I think we're good to merge.

taras commented:
@martndemus done

martndemus commented:
@taras Squash your commits please!

poteto commented:
You can squash PRs in Github @martndemus. I'm outside right now so can't make a release

martndemus commented:
Published `v0.27.0`

taras commented:
Awesome. Thank you
